ROADMAP.md

```diff
@@ -0,0 +1,277 @@
+downstream users to easier integrate with rust-libp2p, I think this counts as a "improving existing
+components" over "introducing a new component".
+
+First draft is in https://github.com/libp2p/rust-libp2p/pull/2828
+
+## Kademlia efficient querying
+
+Status: in progress
+
+Tracking: https://github.com/libp2p/rust-libp2p/pull/2712
+
+Dependencies:
+
+Dependents:
+
+Users of rust-libp2p like [iroh](https://github.com/n0-computer/iroh) need this for low latency usage of `libp2p-kad`.
+
+## Kademlia client mode
+
+Status: todo
+
+Tracking: https://github.com/libp2p/rust-libp2p/issues/2032
+
+Dependencies:
+
+- [Cross behaviour communication](#cross-behaviour-communication)
+
+Dependents:
+
+## Hole punching
+
+Status: todo
+
+Dependencies:
+
+Dependents:
+
+We released hole punching support with [rust-libp2p
+`v0.43.0`](https://github.com/libp2p/rust-libp2p/releases/tag/v0.43.0), see also
+https://github.com/libp2p/rust-libp2p/issues/2052. We are currently collecting data via the
+[punchr](https://github.com/dennis-tra/punchr) project on the hole punching success rate. See also
+[call for
+action](https://discuss.libp2p.io/t/decentralized-nat-hole-punching-measurement-campaign/1616) in
+case you want to help. Based on this data we will likely find many optimizations we can do to our
+hole punching stack.
+
+## Streaming response protocol aka. the ultimate escape hatch
+
+Status: todo
+
+Tracking: https://github.com/libp2p/rust-libp2p/issues/2657
+
+rust-libp2p is very opinionated on how to write peer-to-peer protocols. There are many good reasons
+for this, and I think we should not change directions here. That said, the ultimate escape hatch -
+allowing users to create a stream and do whatever they want with it - will make it easier for
+newcomers to get started.
+
+## Improved WASM support
+
+Status: todo
+
+Tracking: https://github.com/libp2p/rust-libp2p/issues/2617
+
+Dependencies:
+
+Dependents:
+
+- WebRTC browser-to-browser and browser side
+
+This opens rust-libp2p to hole new set of use-cases.
+
+## Muxer handshake optimization
+
+Status: todo
+
+Tracking: https://github.com/libp2p/rust-libp2p/issues/2994
+
+I suggest we do not prioritize this. On the performance end, investing into QUIC support seems like
+the better bet. Long term this is a wonderful improvement, likely very much appreciated by users
+though not thus far demanded.
+
+## Bitswap implementation
+
+Status: todo
+
+Tracking: https://github.com/libp2p/rust-libp2p/issues/2632
+
+I think this is a common component that many users need to build peer-to-peer applications. In
+addition, it is very performance critical and thus likely challenges many of our existing designs in
+rust-libp2p.
+
+I would prioritize it below [Muxer handshake optimization](#muxer-handshake-optimization) following
+the convention of improving existing components over introducing new ones. Users have and can
+implement their own implementations and are thus not blocked on the rust-libp2p project.
```

Comment:
Why included these here if they are low priority?

Reply:
I consider all items on this roadmap worth doing eventually. Given that we are resource constraint we have to make trade offs somewhere. That said, I do think it is worth tracking them here for future planning.

I suggest we move this discussion to https://github.com/libp2p/rust-libp2p/pull/2997#discussion_r989168789. I am resolving here.